clippy_utils/src/eager_or_lazy.rs

```diff
@@ -9,128 +9,191 @@
+        cmp::max(self, rhs)
     }
 }
-
-/// Identify some potentially computationally expensive patterns.
-/// This function is named so to stress that its implementation is non-exhaustive.
-/// It returns FNs and FPs.
-fn identify_some_potentially_expensive_patterns<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) -> bool {
-    // Searches an expression for method calls or function calls that aren't ctors
-    struct FunCallFinder<'a, 'tcx> {
-        cx: &'a LateContext<'tcx>,
-        found: bool,
+impl ops::BitOrAssign for EagernessSuggestion {
+    fn bitor_assign(&mut self, rhs: Self) {
+        *self = *self | rhs;
     }
+}
 
-    impl<'a, 'tcx> intravisit::Visitor<'tcx> for FunCallFinder<'a, 'tcx> {
-        type Map = Map<'tcx>;
+/// Determine the eagerness of the given function call.
+fn fn_eagerness(cx: &LateContext<'tcx>, fn_id: DefId, name: Symbol, args: &'tcx [Expr<'_>]) -> EagernessSuggestion {
+    use EagernessSuggestion::{Eager, Lazy, NoChange};
+    let name = &*name.as_str();
 
-        fn visit_expr(&mut self, expr: &'tcx Expr<'tcx>) {
-            let call_found = match &expr.kind {
-                // ignore enum and struct constructors
-                ExprKind::Call(..) => !is_ctor_or_promotable_const_function(self.cx, expr),
-                ExprKind::Index(obj, _) => {
-                    let ty = self.cx.typeck_results().expr_ty(obj);
-                    is_type_diagnostic_item(self.cx, ty, sym::hashmap_type)
-                        || is_type_diagnostic_item(self.cx, ty, sym::BTreeMap)
-                },
-                ExprKind::MethodCall(..) => true,
-                _ => false,
-            };
+    let ty = match cx.tcx.impl_of_method(fn_id) {
+        Some(id) => cx.tcx.type_of(id),
+        None => return Lazy,
+    };
 
-            if call_found {
-                self.found |= true;
-            }
-
-            if !self.found {
-                intravisit::walk_expr(self, expr);
+    if (name.starts_with("as_") || name == "len" || name == "is_empty") && args.len() == 1 {
+        if matches!(
+            cx.tcx.crate_name(fn_id.krate),
+            sym::std | sym::core | sym::alloc | sym::proc_macro
+        ) {
+            Eager
+        } else {
+            NoChange
+        }
+    } else if let Some(ty::Adt(def, subs)) = cx.tcx.impl_of_method(fn_id).map(|id| cx.tcx.type_of(id).kind()) {
+        // Types where the only fields are generic types (or references to) with no trait bounds other
+        // than marker traits.
+        // Due to the limited operations on these types functions should be fairly cheap.
+        if def
+            .variants
+            .iter()
+            .flat_map(|v| v.fields.iter())
+            .any(|x| matches!(cx.tcx.type_of(x.did).peel_refs().kind(), ty::Param(_)))
+            && all_predicates_of(cx.tcx, fn_id)
+                .trait_predicates()
+                .all(|(pred, _)| cx.tcx.trait_def(pred.skip_binder().trait_ref.def_id).is_marker)
```

Comment:
Okay, I think I found a way to simplify this

```rust
cx.tcx.predicates_of(fn_id)
    .instantiate_identity(cx.tcx)
    .predicates
    .into_iter()
    .all(|p| match p.kind().skip_binder() {
        ty::PredicateKind::Trait(p) => cx.tcx.trait_def(p.def_id()).is_marker,
        _ => true,
    })
```

Reply:
Or this :)

```rust
std::iter::successors(Some(cx.tcx.predicates_of(fn_id)), |p| {
    p.parent.map(|p| cx.tcx.predicates_of(p))
})
.flat_map(|p| p.predicates)
.all(|(p, _)| match p.kind().skip_binder() {
    ty::PredicateKind::Trait(p) => cx.tcx.trait_def(p.def_id()).is_marker,
    _ => true,
});
```

Reply:
Removed the new iterator, but kept the function.